### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by fepitre**
https://github.com/QubesOS/qubes-linux-kernel/pull/284 and https://github.com/QubesOS/qubes-linux-kernel/pull/285 are now needed.

### CODE REVIEW

**Comment by fepitre** [comment]
I've just built python3-pyudev for CentOS 8.

File: Makefile [hunk 1]
```diff
@@ -3,13 +3,13 @@ VERSION := $(shell cat version 2>/dev/null)
 KVER ?= $(shell uname -r)
 KVER_MAJOR ?= $(word 1, $(subst ., ,$(KVER))).$(word 2, $(subst ., ,$(KVER)))
 
-# BEGIN LOCAL BUILD PART
-obj-m += module/dummy_psu.o
-
 ifeq (1,$(DEBUG))
-CFLAGS_dummy_psu.o := -DDEBUG
+export CFLAGS ?= -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
```
Comment (marmarek) [inline Makefile:7]: Why hardening flags are only in debug build?

File: module/dummy_psu.c [hunk 2]
```diff
@@ -9,9 +9,390 @@
...
+	int cr;
+
+	strncpy(buf, key, MAX_KEYLENGTH);
+	buf[MAX_KEYLENGTH - 1] = '\0';
+
+	cr = strnlen(buf, MAX_KEYLENGTH) - 1;
+	if (cr < 0)
+		return -1;
+	if (buf[cr] == '\n')
+		buf[cr] = '\0';
```
Comment (marmarek) [inline module/dummy_psu.c:385]: In what case you get `\n` appended to the property name?

File: module/dummy_psu.c [hunk 3]
```diff
@@ -135,58 +519,73 @@ long dummy_psu_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
...
 			if (pobj->num_supplicants >= MAX_SUPPLICANTS) {
-				pr_err("%s: Cannot add supplicant. Number of supplicants is reached\n",
+				pr_err("IOCTL_PSU_ADD_SUPPLIED_TO: %s: Cannot add supplicant. Number of supplicants is reached\n",
 				       pobj->dev_name);
 				return -EFAULT;
 			}
+			if (copy_from_user(&supplied_to, (void __user *)arg,
+					   sizeof(char[MAX_KEYLENGTH])) != 0)
+				return -EFAULT;
 			pobj->supplied_to[pobj->num_supplicants] = supplied_to;
```
Comment (marmarek) [inline module/dummy_psu.c]: `supplied_to` is a local variable, you can't  use it in long-lived struct. You need strdup or similar, but make sure it is 0-terminated first.

File: receiver/qubes-psu-client.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		jobj = json_tokener_parse(buf);
+
+		json_object_object_get_ex(jobj, "NAME", &tmp);
+		strncpy(curr_name, json_object_get_string(tmp), MAX_KEYLENGTH - 1);
+
+		json_object_object_get_ex(jobj, "TYPE", &tmp);
+		strncpy(curr_type, json_object_get_string(tmp), MAX_KEYLENGTH - 1);
+
+		dev_num = -1;
+		if (strcmp(curr_name, "\0") && strcmp(curr_name, "\0")) {
```
Comment (marmarek) [inline receiver/qubes-psu-client.c]: This is the same as `strcmp(..., "")`, and also - the second one should be curr_type.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (fepitre) [reply Makefile:7]: Normally it's RPM who gives them not the Makefile.

File: module/dummy_psu.c (hunk 2 above)
Comment (fepitre) [reply module/dummy_psu.c:385]: I've cherry-picked it from `test_power.c` (specified few lines above in comment) but this case is not verified but possible.

**Review by marmarek — APPROVED** [review]

**Pull request merged**